basic keystone tests

thutterer opened the pull request:
@jsuchome This is how far i got. 
@vmoravec This is not meant to be merged yet! Lot of typos and such. But a short feedback on how I implemented the two tests would still be appreciated.

thutterer committed:
add some basic Keystone tests

aspiers commented on features/keystone.feature:
```diff
@@ -0,0 +1,15 @@
+
+  Scenario: User create # was test no.30
+    Given I have a tenant named 'openstack'
+    When I can delete a user named 'test_user'
```
Just a grammar nitpick: this should read "When I delete ...", because a "When $X" clause is not concerned with whether the $X action is possible - it's part of the test setup, so it already assumes it _must_ be possible.  In contrast, "Then" clauses _are_ concerned with stating that the action is possible, so "Then I can ..." is correct.

aspiers commented on features/step_definitions/keystone/users_steps.rb:
```diff
@@ -0,0 +1,40 @@
+  expect(@tenants).to include @tenant
+end
+
+Given(/^I can delete a user named 'test_user'$/) do
```
```
Given(/^I (can )?delete a user named 'test_user'$/) do
```

That means it can be used in both "When ..." and "Then ..." clauses.

vmoravec commented on config/features/keystone.yml:
```diff
@@ -0,0 +1,2 @@
+  user_name: test_user
+  tenant_name: openstack
```
Just don't forget to fix the yaml file content ;)

``` yaml
features:
  keystone:
    scenario_tag:
```

thutterer committed:
add commands to openstack client class

thutterer commented:
I made some progress but it's still not polished.
The following is still missing:
- a tag for each scenario ... we all agreed on that, right? or is there a different approach?
- separate scenario for testing deletion of user ... so that it'll be ok to implicitly doing it in a hook after each scenario.
- better step and scenario description ... feel free to propose some.

As you may have realized by now: I tried to follow @aspiers' proposals in #2 while @jsuchome seems to have used more of @vmoravec style. So maybe we can gain a better comparison now between both PRs and approaches :)

jsuchome commented on features/users.feature:
```diff
@@ -0,0 +1,22 @@
+
+  Background:
+    Given project 'openstack' exists
+    And user 'test_user' exists
```
So you are creating user in the Background? I understand it is something needed for all scenarios, but creation of user is something that we are testing, so shouldn't it be rather part of test scenario itself?

thutterer replied:
> So you are creating user in the Background? I understand it is something needed for all scenarios, but creation of user is something that we are testing, so shouldn't it be rather part of test scenario itself?

Probably the same as with deletion: I could add a dummy scenario for user creation that does nothing but the background.
Edit: But imho it is visible enough as it is.

thutterer commented on features/step_definitions/users/background_steps.rb:
```diff
@@ -0,0 +1,46 @@
+      :project  => @project_name
+    control_node.openstack.role.add "Member",
+      :project  => @project_name,
+      :user     => @user_name
```
Should this (role add) better be in a separate scenario?

thutterer commented on lib/cct/commands/openstack.rb:
```diff
@@ -69,15 +74,29 @@ def create name, options={}
           params.clear
-          user_row = Struct.new(:id, :name)
-          result = exec!("list", "--format=csv").output
+          user_row = Struct.new(:id, :name, :subnets)
```
@vmoravec Maybe it would be better to rewrite this as a OpenStruct. I guess `network` won't be the only component that has a third (or fourth) column.

vmoravec replied:
I would rather go with checking the last item of `options` for hash and search for `:row` param which might provide the exact struct as needed.  Something like:

``` ruby
params = options.last.is_a?(Hash) ? options.pop : {}
row = params[:row] || Struct.new(:id, :name)
```

It's better to keep the parent class as much generic as possible.

vmoravec replied:
And the `Openstack::Network` would implement it's own `list` method that would call `super(options << { row: Struct.new(:id, :name, :subnets)} )` or something like this.

thutterer commented on lib/cct/commands/openstack.rb:
```diff
@@ -69,15 +74,29 @@ def create name, options={}
         end
 
-        def list
+        def list *options
```
@vmoravec That might be a bit hacky but I needed a simple way to add commandline options that you haven't provided.

thutterer commented on tasks/features/users.rake:
```diff
@@ -0,0 +1,13 @@
+
+  namespace :users do
+    desc "First feature task"
+    feature_task :first, tags: :@first
```
Oh I forgot that file. I'll add tagged feature tasks later.

toabctl commented:
From a first view it seems that you are currently reinventing tempest. I thought this testsuite is basically for barclamps, crowbar and the overall product. We are already running scenario tests for the openstack components. Imho there's no need to rewrite tests here. If something is missing, add it to tempest.
Or am I missing something? I though this works (somehow simplified) like this:
- a scenario configures our product and all the barclamps (i.e. different combinations of network configurations and cinder backends)
- then it runs tempest (or a subset of tempest) to verify that openstack is still working

Is that assumption wrong?

vmoravec commented:
I agree that this and other PRs overlap partly with what tempest does, but I think it's important to have basic tests in our testsuite before adding more advanced test cases. I guess this was part of the reasoning why we started with porting the tests from [qa-openstack-cli repo](https://github.com/SUSE-Cloud/qa-openstack-cli) . If I'm mistaken, please correct me.

Another reason for having such tests  might be that we want to test the official openstack client api as tempest does not do that. Tempest docs say:  `API tests should always use the Tempest implementation of the OpenStack API, as we want to ensure that bugs aren't hidden by the official clients.`

Regarding the comitting of tempest updates to upstream, I'm not sure how that works for us as we actually use an outdated tempest version. Would upstream be ready to accept updates to an older tempest release? I don't know how is this managed in upstream at all. @kbaikov do you know more on that, please?

thutterer commented:
Updated! Now with tags, some rewording of scenario descriptions and a scenario for deletion.

aspiers commented:
Looking really good! :-)

However, I think you should drop the Rake stuff entirely, including the scenario tags.   [Upstream recommends avoiding running Cucumber via `rake`](https://github.com/cucumber/cucumber/wiki/using-rake), and [the `-n` option already gives you the same functionality and much more](https://www.relishapp.com/cucumber/cucumber/docs/cli/run-feature-elements-matching-a-name-with-name-n).

aspiers commented:
@vmoravec I agree with you - I think some overlap with Tempest is OK, especially in the beginning.

thutterer commented:
I've amended my 2nd commit another time:
- Some minor ruby refactoring.
- Changed the step "Given I create a project" to "Given project 'openstack' exists" because I realized this project is created by Chef. When I deleted it manually and ran my tests (to test the creation code part), it was not possible for user Admin to authenticate until chef-client was run and recreated project openstack. So it made no sense in my scenario to try to create it as this would fail anyway.

toabctl commented:
> I agree that this and other PRs overlap partly with what tempest does, but I think it's important to have basic tests in our testsuite before adding more advanced test cases. I guess this was part of the reasoning why we started with porting the tests from qa-openstack-cli repo . If I'm mistaken, please correct me.

Ok.

> Another reason for having such tests might be that we want to test the official openstack client api as tempest does not do that. Tempest docs say: API tests should always use the Tempest implementation of the OpenStack API, as we want to ensure that bugs aren't hidden by the official clients.

I thought there are functional tests in the client code which can be executed. But looking into some of the clients, there's not much available. And afaics you are not only testing the client API but also the client cli code.

> Regarding the comitting of tempest updates to upstream, I'm not sure how that works for us as we actually use an outdated tempest version. Would upstream be ready to accept updates to an older tempest release? I don't know how is this managed in upstream at all. @kbaikov do you know more on that, please?

I think the reason why we are running an outdated version is that at some point after a release we can no longer satisfy the needed dependencies. But you could run tempest in a virtualenv from git or install the packages from Cloud:OpenStack:Master on any host that has access to the cloud.

vmoravec commented on features/support/env.rb:
```diff
@@ -42,6 +42,15 @@
+    control_node.openstack.user.delete user_name
+  end
+end
+
```
Please don't use tagged hooks. We are mixing here too specific implementation details from the `users` feature with high level code. 
This could be done more transparently in `Background` with `And the test user does not exists"` right after the first `Given`.

thutterer replied:
> This could be done more transparently in Background with And the test user does not exists" right after the first Given.

But wouldn't that mean that the cleaning-up part is not executed _after_ all tests are finished. So test_user would still exist then and only get deleted at a rerun?

aspiers replied:
This is exactly the same discussion we already had at length in #2, except this time it's about users instead of images.

> But wouldn't that mean that the cleaning-up part is not executed after all tests are finished. So test_user would still exist then and only get deleted at a rerun?

Yes, correct.  I don't see a problem with that approach in general; in fact it's quite a nice trick and I'm glad @vmoravec discovered it in the upstream Cucumber community.  For images you might risk exceeding quotas, but again the quota usage can be checked in the background.

As a general principle it seems that unlike @vmoravec, I think it can be fine to use tagged hooks.  However in this case I think I agree with him, because the tag is too broad to have a user deletion hook attached to it.  A more suitable tag name would be `@cleanup_users` or similar.  A `@users` tag should be used to tag _all_ scenarios which involve testing user management of any sort, and not all of those scenarios would create a new user and consequently require user deletion.

vmoravec commented on features/support/env.rb:
```diff
@@ -42,6 +42,15 @@
+  user_names = users.map(&:name)
+  if user_names.include? user_name
+    control_node.openstack.user.delete user_name
+  end
```
This might be written shorter by using `find`:

``` ruby
if control_node.openstack.user.list.find {|u| u.name == user_name }
  control_node.openstack.user.delete(user_name)
end
```

thutterer commented:
ok, amended one more time. the cleaning up part is now in the background. no more after-hooks :)
also the code for checking for user-existance is now a bit shorter.

vmoravec commented:
@thutterer It looks good, I tested it and works fine. Thanks!
I will improve the customization of the `list` method later.
